design/velero-upgrade.md

```diff
@@ -0,0 +1,80 @@
+For an concrete example, in v1.10, we have done some refactors to Velero's pod volume backup (a.k.a Restic backup formerly), two of the refactors are:  
+- Rename some Velero server and daemonset parameters
+- Rename Velero daemonset from "restic" to "node-agent"
+
+As a result, the existing upgrade approach is not sufficient:
+- If we still use manual steps, we have to tell users the detailed parameters that have been renamed, and users then needs to check their existence and modify them one by one.
+- If we still use manual steps, users have to delete the old restic daemonset and run ```velero install``` with the full parameters they specified during the fresh installation, because for one thing, Kubernetes doesn't support to rename a resource; for another, Velero doesn't provide a way to install a single resource only. 
+
+## Solution
+
+We want to add a new command besides ```velero install``` called ```velero upgrade```. In a upgrade case, users could call ```velero upgrade``` without bothering to know the details. This bring below benefits:  
+- Solve the upgrade problems in v1.10
+- Make Velero more user friendly, because the upgrade process is significantly simplified
+- Support future evolvements. For example, in future, Velero will have various data movement features, for which we will add new moudles inevitably. Then we can add enhancement incrementally to ```velero upgrade```, because inside the command, we can do whatever complex things 
+
+The ```velero upgrade``` command will fully replace the existing manual steps in the aforementioned Velero document.
+
+## Design
+
+At present, ```velero upgrade``` will do below things:
+- Get the new version from the client binary itself, that is, the upgrade target version
+- Get the old version from the existing velero server, that is, the current version
+- Retrieve Velero server deployment object
+- Retrieve Velero daemonset object
+- Iterate from current version through target version, apply the changes (i.e., the image, name and parameter changes) to the deployment object once per version. For example, when upgrading from 1.9 to 1.11, we need to apply the changes between 1.9 and 1.10 first, then apply the changes between 1.10 and 1.11. See the details below for how to drive the changes apply
+- Do the same changes apply to Velero daemonset object as the deployment object
+- Update the Velero server deployment from the modified deployment object
+- Delete the existing Velero daemonset
+- Create the new Velero daemonset from the modified daemonset object
+- Extract the new CRDs from the client binary and apply them
+
+### Action Table
+We create below action functions to apply changes for each target version, the number in the function name indicates the target version:  
```

Comment:
I think there are more details we need to be careful with, for example, if we released 1.10.0 then 1.11.0, then 1.10.1 then 1.11.1.

How do we manage the upgrade path if 1.11.1 require user to upgrade to 1.10.1 but 1.10.1 cannot be upgraded to 1.11.0?

Do we need to add more pre-requisite items?

Reply:
First of all, if we have this situation -- "1.10.1 cannot be upgraded to 1.11.0", this means Velero have multiple side streams, it seems we don't have this behavior at present? Moreover, in order to support side streams, our entire release cycle management, code management, and team resource management have to be changed, so do we need to consider this in the current design?

Secondly, once Velero has this multiple stream manner, it doesn't mean this action table concept not work totally, merely, the action table will not be a linear vector but a 2d tree like structure. Therefore, can we add this requirement to the NonGoal section and consider it later?

Finally, if we only consider the single stream release module only, this action table concept could still support patch releases. On the other hand, as a principle, the patch releases don't include complex changes or user experience changes, we only need to provide a general action function, that is to change the image only, but this is not mandatory.

Reply:
Let's further discuss on this then I will make the change to the design.